examples/energy-management-app/energy-management-common/src/DeviceEnergyManagementDelegateImpl.cpp

```diff
@@ -47,44 +79,246 @@ using CostsList = DataModel::List<const Structs::CostStruct::Type>;
  *   6) generate a PowerAdjustEnd event with cause NormalCompletion
  *   7) if necessary, update the forecast with new expected end time
  */
-Status DeviceEnergyManagementDelegate::PowerAdjustRequest(const int64_t power, const uint32_t duration, AdjustmentCauseEnum cause)
+Status DeviceEnergyManagementDelegate::PowerAdjustRequest(const int64_t powerMw, const uint32_t durationS,
+                                                          AdjustmentCauseEnum cause)
 {
-    Status status = Status::UnsupportedCommand; // Status::Success;
+    bool generateEvent = false;
 
-    // TODO: implement
-    mEsaState = ESAStateEnum::kPowerAdjustActive;
+    // If a timer is running, cancel it so we can start it with the new duration
+    if (mPowerAdjustmentInProgress)
+    {
+        DeviceLayer::SystemLayer().CancelTimer(PowerAdjustTimerExpiry, this);
+    }
+    else
+    {
+        // Going to start a new power adjustment so will need to generate an event
+        generateEvent = true;
+
+        // Record when this PowerAdjustment starts. Note if we do not set this value if a PowerAdjustment is in progress
+        CHIP_ERROR err = GetEpochTS(mPowerAdjustmentStartTimeUtc);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(AppServer, "Unable to get time: %" CHIP_ERROR_FORMAT, err.Format());
+            return Status::Failure;
+        }
+    }
 
-    // TODO:  Generate a PowerAdjustStart Event, then begins to adjust its power
-    // When done, raise PowerAdjustEnd & ESAState set to kOnline.
+    //  Update the forecast with the new expected end time
+    if (mpDEMManufacturerDelegate != nullptr)
+    {
+        CHIP_ERROR err = mpDEMManufacturerDelegate->HandleDeviceEnergyManagementPowerAdjustRequest(powerMw, durationS, cause);
+        if (err != CHIP_NO_ERROR)
+        {
+            return Status::Failure;
+        }
+    }
 
-    MatterReportingAttributeChangeCallback(mEndpointId, DeviceEnergyManagement::Id, ESAState::Id);
+    SetESAState(ESAStateEnum::kPowerAdjustActive);
 
-    return status;
+    switch (cause)
+    {
+    case AdjustmentCauseEnum::kLocalOptimization:
+        mPowerAdjustCapabilityStruct.Value().cause = PowerAdjustReasonEnum::kLocalOptimizationAdjustment;
```

Comment:
Again, like yesterday:



What guarantees that mPowerAdjustCapabilityStruct has a value here? Should probably be documented.

Same in the other cases in this switch.

Reply:
I have documented why it is guaranteed to have a value.